File: tools/interp/pipeline.py
"""Generate the 60 fps interpolated frames for one capture.

The run has four parts. It composes every pose on one padded canvas, writes the
colour and alpha pictures RIFE reads, runs RIFE once per channel over the whole
capture, then divides the colour back out and writes RGBA frames plus a
manifest.

RIFE's directory mode maps output i to input position i * count / numframe and
interpolates between the two poses around it. Picking numframe so that ratio
equals one 60 fps step measured in poses puts every output exactly on a sample
time, which is why the whole capture takes two RIFE processes instead of one per
frame. The pose list is padded with copies of the last pose until the ratio is
exact; the padding only feeds output positions past the last real pose, which
the hold covers with a copy instead.
"""

import hashlib
import json
import multiprocessing
import os
import shutil
import subprocess
import sys
import time
from fractions import Fraction

from . import fixture, images, png, timeline

MODEL_NAME = "rife-v4.6"
DEFAULT_BINARY = ".cache/rife/bin/rife-ncnn-vulkan"
DEFAULT_MODEL = ".cache/rife/bin/rife-v4.6"


def sha256(path):
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for block in iter(lambda: fh.read(1 << 20), b""):
            h.update(block)
    return h.hexdigest()


def tool_commit(root):
    """Return the git commit of the repository, or None outside git."""
    try:
        out = subprocess.run(["git", "-C", root, "rev-parse", "HEAD"],
                             capture_output=True, text=True, timeout=10)
    except (OSError, subprocess.SubprocessError):
        return None
    return out.stdout.strip() or None if out.returncode == 0 else None


def load_spec(assets, capture, resolved_path):
    """Read resolved.json, or build the fixture when it is not there yet."""
    path = resolved_path or os.path.join(assets, "captures", capture, "resolved.json")
    if os.path.exists(path):
        with open(path) as fh:
            return json.load(fh), path, sha256(path)
    spec = fixture.build(capture, "original/win3x/cd")
    return spec, None, None


def _decimal(value):
    """Round a Fraction of milliseconds to six decimal places as a float."""
    scaled = value * 1000000
    whole = (scaled.numerator * 2 + scaled.denominator) // (scaled.denominator * 2)
    return whole / 1000000


def prepare(spec, source_dir, work_dir):
    """Compose every pose and write the colour and alpha pictures for RIFE.

    Returns the canvas rectangle and the composed RGBA buffers.
    """
    rect = images.union_rect(spec["poses"])
    _, _, width, height = rect
    rgb_dir = os.path.join(work_dir, "rgb")
    alpha_dir = os.path.join(work_dir, "alpha")
    for path in (rgb_dir, alpha_dir):
        shutil.rmtree(path, ignore_errors=True)
        os.makedirs(path)
    composed = []
    for i, pose in enumerate(spec["poses"]):
        rgba = images.compose(os.path.join(source_dir, pose["image"]), pose, rect)
        composed.append(rgba)
        rgb, alpha = images.split(rgba)
        png.write_rgb(os.path.join(rgb_dir, "%05d.png" % i), width, height,
                      images.rows(rgb, width, 3), level=1)
        png.write_gray(os.path.join(alpha_dir, "%05d.png" % i), width, height,
                       images.rows(alpha, width, 1), level=1)
    return rect, composed


def pad_inputs(work_dir, count, multiple):
    """Copy the last pose picture until the input count divides evenly."""
    padded = count
    while padded % multiple:
        for name in ("rgb", "alpha"):
            src = os.path.join(work_dir, name, "%05d.png" % (count - 1))
            shutil.copyfile(src, os.path.join(work_dir, name, "%05d.png" % padded))
        padded += 1
    return padded


def run_rife(binary, model, in_dir, out_dir, numframe, gpu=None):
    """Run one RIFE pass over a directory and return the command line."""
    shutil.rmtree(out_dir, ignore_errors=True)
    os.makedirs(out_dir)
    cmd = [binary, "-m", model, "-i", in_dir, "-o", out_dir, "-n", str(numframe)]
    if gpu is not None:
        cmd += ["-g", str(gpu)]
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError("rife failed: %s\n%s" % (" ".join(cmd), result.stderr[-2000:]))
    return cmd


def run_rife_pair(binary, model, first, second, out_path, s, gpu=None):
    """Run one RIFE inference for a single time step."""
    cmd = [binary, "-m", model, "-0", first, "-1", second,
           "-s", "%.9f" % s, "-o", out_path]
    if gpu is not None:
        cmd += ["-g", str(gpu)]
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError("rife failed: %s\n%s" % (" ".join(cmd), result.stderr[-2000:]))
    return cmd


def _recombine_one(job):
    """Worker: read one colour and alpha frame, write one RGBA frame."""
    rgb_path, alpha_path, out_path, width, height, cutoff = job
    rw, rh, rc, rgb = png.read(rgb_path)
    aw, ah, ac, alpha = png.read(alpha_path)
    if (rw, rh) != (width, height) or (aw, ah) != (width, height):
        raise ValueError("rife returned %dx%d for %s" % (rw, rh, rgb_path))
    if rc != 3:
        rgb = bytes(b for i, b in enumerate(rgb) if i % rc < 3)
    grey = bytes(alpha[i * ac] for i in range(width * height)) if ac != 1 else bytes(alpha)
    rgba = images.combine(bytes(rgb), grey, cutoff)
    png.write_rgba(out_path, width, height, images.rows(rgba, width, 4))
    return out_path


def _write_copy(job):
    """Worker: write one source copy frame."""
    out_path, width, height, rgba = job
    png.write_rgba(out_path, width, height, images.rows(rgba, width, 4))
    return out_path


def generate(capture, assets, out_dir, fps=60, resolved_path=None,
             binary=DEFAULT_BINARY, model=DEFAULT_MODEL, dry_run=False,
             jobs=None, keep_work=False):
    """Build every output frame and the manifest. Returns the manifest."""
    started = time.time()
    root = os.getcwd()
    spec, spec_path, spec_hash = load_spec(assets, capture, resolved_path)
    source_dir = os.path.join(assets, "captures", capture)
    entries = timeline.plan(spec, fps)
    gap = timeline.uniform_gap(spec)
    step = Fraction(1000, fps) / gap if gap else None
    batched = step is not None and step < 1

    if dry_run:
        kinds = {}
        for entry in entries:
            kinds[entry["kind"]] = kinds.get(entry["kind"], 0) + 1
        rect = images.union_rect(spec["poses"])
        print("capture %s, %d poses, end %d ms, %d frames at %d fps"
              % (capture, len(spec["poses"]), spec["end_ms"], len(entries), fps))
        print("canvas %dx%d at (%d, %d)" % (rect[2], rect[3], rect[0], rect[1]))
        print("frames by kind: %s" % ", ".join("%s %d" % kv for kv in sorted(kinds.items())))
        print("batched rife: %s" % ("yes, step %s poses" % step if batched else "no"))
        return None

    os.makedirs(out_dir, exist_ok=True)
    work_dir = os.path.join(out_dir, "work")
    os.makedirs(work_dir, exist_ok=True)
    rect, composed = prepare(spec, source_dir, work_dir)
    _, _, width, height = rect

    commands = []
    frame_paths = {}
    if batched:
        padded = pad_inputs(work_dir, len(spec["poses"]), step.numerator)
        numframe = int(padded / step)
        for name in ("rgb", "alpha"):
            commands.append(run_rife(binary, model,
                                     os.path.join(work_dir, name),
                                     os.path.join(work_dir, name + "_out"),
                                     numframe))
        for n, entry in enumerate(entries):
            if entry["kind"] == timeline.INTERP:
                frame_paths[n] = (
                    os.path.join(work_dir, "rgb_out", "%08d.png" % (n + 1)),
                    os.path.join(work_dir, "alpha_out", "%08d.png" % (n + 1)),
                )
    else:
        for name in ("rgb_out", "alpha_out"):
            path = os.path.join(work_dir, name)
            shutil.rmtree(path, ignore_errors=True)
            os.makedirs(path)
        for n, entry in enumerate(entries):
            if entry["kind"] != timeline.INTERP:
                continue
            k = entry["pose"]
            pair = []
            for name in ("rgb", "alpha"):
                out_path = os.path.join(work_dir, name + "_out", "%08d.png" % (n + 1))
                commands.append(run_rife_pair(
                    binary, model,
                    os.path.join(work_dir, name, "%05d.png" % k),
                    os.path.join(work_dir, name, "%05d.png" % (k + 1)),
                    out_path, float(entry["s"])))
                pair.append(out_path)
            frame_paths[n] = tuple(pair)

    index = {p["index"]: i for i, p in enumerate(spec["poses"])}
    copy_jobs = []
    interp_jobs = []
    names = []
    for n, entry in enumerate(entries):
        name = "frame%05d.png" % n
        names.append(name)
        out_path = os.path.join(out_dir, name)
        if entry["kind"] == timeline.INTERP:
            rgb_path, alpha_path = frame_paths[n]
            interp_jobs.append((rgb_path, alpha_path, out_path, width, height,
                                images.ALPHA_CUTOFF))
        else:
            copy_jobs.append((out_path, width, height,
                              bytes(composed[index[entry["source"][0]]])))

    with multiprocessing.Pool(jobs or os.cpu_count()) as pool:
        for _ in pool.imap_unordered(_write_copy, copy_jobs, chunksize=4):
            pass
        for _ in pool.imap_unordered(_recombine_one, interp_jobs, chunksize=4):
            pass

    manifest = build_manifest(capture, spec, spec_path, spec_hash, entries, names,
                              rect, fps, binary, model, commands, source_dir, root)
    manifest["wall_seconds"] = round(time.time() - started, 2)
    with open(os.path.join(out_dir, "manifest.json"), "w") as fh:
        json.dump(manifest, fh, indent=1)
        fh.write("\n")
    if not keep_work:
        shutil.rmtree(work_dir, ignore_errors=True)
    return manifest


def build_manifest(capture, spec, spec_path, spec_hash, entries, names, rect, fps,
                   binary, model, commands, source_dir, root):
    """Describe the run in enough detail to reproduce and check it."""
    x, y, width, height = rect
    frames = []
    for entry, name in zip(entries, names):
        frame = {
            "file": name,
            "t_ms": _decimal(entry["t"]),
            "duration_ms": _decimal(entry["duration"]),
            "kind": entry["kind"],
            "source": entry["source"],
            "s": float(entry["s"]) if entry["s"] is not None else None,
        }
        frames.append(frame)
    sounds = [{"pose": p["index"], "t_ms": p["t_ms"], "sound": p["sound"]}
              for p in spec["poses"] if p.get("sound")]
    return {
        "capture": capture,
        "fps": fps,
        "tool": {
            "commit": tool_commit(root),
            "model": MODEL_NAME,
            "model_sha256": sha256(os.path.join(model, "flownet.bin")),
            "binary_sha256": sha256(binary),
            "alpha_cutoff": images.ALPHA_CUTOFF,
            "commands": [" ".join(c) for c in commands[:4]],
            "command_count": len(commands),
        },
        "input": {
            "resolved": spec_path,
            "resolved_sha256": spec_hash,
            "source_dir": source_dir,
            "from_fixture": spec_path is None,
            "poses": [
                {"index": p["index"], "t_ms": p["t_ms"], "image": p["image"],
                 "x": p["x"], "y": p["y"], "w": p["w"], "h": p["h"],
                 "sha256": sha256(os.path.join(source_dir, p["image"]))}
                for p in spec["poses"]
            ],
        },
        "canvas": spec["canvas"],
        "frame_rect": {"x": x, "y": y, "w": width, "h": height},
        "end_ms": spec["end_ms"],
        "final_wav": spec.get("final_wav"),
        "cuts": spec.get("cuts") or [],
        "sounds": sounds,
        "frames": frames,
    }

File: src/interp/pipeline.cpp
#include "interp/pipeline.h"

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <functional>
#include <iostream>
#include <map>
#include <mutex>
#include <numeric>
#include <stdexcept>
#include <thread>
#include <unistd.h>
#include <vector>

#include "export/png_write.h"
#include "export/sha256.h"
#include "interp/images.h"
#include "interp/png_io.h"
#include "interp/process.h"
#include "interp/timeline.h"

namespace swchess::interp {
namespace {

namespace fs = std::filesystem;
using swchess::exporter::encodePng;
using swchess::exporter::kPngGray;
using swchess::exporter::kPngRgb;
using swchess::exporter::kPngRgba;
using swchess::exporter::sha256File;

std::string numbered(const char* pattern, long long value) {
    char buffer[64];
    std::snprintf(buffer, sizeof(buffer), pattern, value);
    return std::string(buffer);
}

void throwIfCancelled(const std::atomic<bool>* cancelled) {
    if (cancelled != nullptr && cancelled->load()) {
        throw std::runtime_error("cancelled");
    }
}

Json readJsonFile(const fs::path& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("cannot read " + path.string());
    }
    return Json::parse(file);
}

void writeFileBytes(const fs::path& path, const std::vector<std::uint8_t>& blob) {
    std::FILE* file = std::fopen(path.c_str(), "wb");
    if (file == nullptr) {
        throw std::runtime_error("cannot write " + path.string());
    }
    const std::size_t written = std::fwrite(blob.data(), 1, blob.size(), file);
    std::fclose(file);
    if (written != blob.size()) {
        throw std::runtime_error("short write on " + path.string());
    }
}

// The commit of the repository the helper is running in, or nothing outside
// a git checkout.
Json toolCommit() {
    try {
        RunResult result = runProgram({"git", "-C", fs::current_path().string(), "rev-parse",
                                       "HEAD"});
        if (result.exitCode != 0) {
            return nullptr;
        }
        std::string text = result.output;
        while (!text.empty() && (text.back() == '\n' || text.back() == '\r' ||
                                 text.back() == ' ')) {
            text.pop_back();
        }
        if (text.empty()) {
            return nullptr;
        }
        return text;
    } catch (const std::exception&) {
        return nullptr;
    }
}

// Where each interpolated sample's colour and alpha picture will land.
struct RifePlan {
    std::vector<std::vector<std::string>> commands;
    std::map<std::size_t, std::pair<std::string, std::string>> paths;
};

RifePlan planRife(const std::vector<PlannedFrame>& frames, const fs::path& workDir,
                  const std::string& binary, const std::string& model) {
    // One list of samples per transition, in the order the frames ask for them.
    std::map<std::size_t, std::vector<std::pair<std::size_t, Fraction>>> wanted;
    for (std::size_t n = 0; n < frames.size(); ++n) {
        if (frames[n].kind == FrameKind::Interp) {
            wanted[frames[n].pose].emplace_back(n, frames[n].step);
        }
    }

    RifePlan plan;
    for (const auto& transition : wanted) {
        const std::size_t k = transition.first;
        std::int64_t denominator = 1;
        for (const auto& sample : transition.second) {
            const std::int64_t other = sample.second.denominator();
            denominator = denominator / std::gcd(denominator, other) * other;
        }

        if (denominator <= kMaxDenominator) {
            const std::string pair = numbered("pair%04lld", static_cast<long long>(k));
            for (const char* channel : {"rgb", "alpha"}) {
                const fs::path pairDir = workDir / channel / pair;
                const fs::path outDir = workDir / (std::string(channel) + "_out") / pair;
                fs::remove_all(pairDir);
                fs::remove_all(outDir);
                fs::create_directories(pairDir);
                fs::create_directories(outDir);
                for (int j = 0; j < 2; ++j) {
                    const fs::path source =
                        workDir / channel /
                        numbered("%05lld.png", static_cast<long long>(k + j));
                    fs::create_hard_link(source, pairDir / (std::to_string(j) + ".png"));
                }
                plan.commands.push_back({binary, "-m", model, "-i", pairDir.string(), "-o",
                                         outDir.string(), "-n",
                                         std::to_string(2 * denominator)});
            }
            for (const auto& sample : transition.second) {
                const std::int64_t step =
                    sample.second.numerator() * (denominator / sample.second.denominator());
                const std::string file = numbered("%08lld.png", static_cast<long long>(step + 1));
                plan.paths[sample.first] = {(workDir / "rgb_out" / pair / file).string(),
                                            (workDir / "alpha_out" / pair / file).string()};
            }
        } else {
            for (const auto& sample : transition.second) {
                std::string outputs[2];
                int slot = 0;
                for (const char* channel : {"rgb", "alpha"}) {
                    const fs::path outDir = workDir / (std::string(channel) + "_out");
                    fs::create_directories(outDir);
                    const fs::path outPath =
                        outDir / numbered("%08lld.png", static_cast<long long>(sample.first));
                    char step[32];
                    std::snprintf(step, sizeof(step), "%.9f", sample.second.toDouble());
                    plan.commands.push_back(
                        {binary, "-m", model, "-0",
                         (workDir / channel / numbered("%05lld.png", static_cast<long long>(k)))
                             .string(),
                         "-1",
                         (workDir / channel /
                          numbered("%05lld.png", static_cast<long long>(k + 1)))
                             .string(),
                         "-s", step, "-o", outPath.string()});
                    outputs[slot++] = outPath.string();
                }
                plan.paths[sample.first] = {outputs[0], outputs[1]};
            }
        }
    }
    return plan;
}

void runCommands(const std::vector<std::vector<std::string>>& commands,
                 const std::atomic<bool>* cancelled) {
    std::atomic<std::size_t> next{0};
    std::mutex lock;
    std::string failure;

    auto worker = [&]() {
        while (true) {
            const std::size_t at = next.fetch_add(1);
            if (at >= commands.size()) {
                return;
            }
            {
                std::lock_guard<std::mutex> guard(lock);
                if (!failure.empty()) {
                    return;
                }
            }
            if (cancelled != nullptr && cancelled->load()) {
                return;
            }
            RunResult result = runProgram(commands[at]);
            if (result.exitCode != 0) {
                std::lock_guard<std::mutex> guard(lock);
                if (failure.empty()) {
                    std::string tail = result.output;
                    if (tail.size() > 2000) {
                        tail = tail.substr(tail.size() - 2000);
                    }
                    failure = "rife failed: " + joinArguments(commands[at]) + "\n" + tail;
                }
                return;
            }
        }
    };

    const int workers = std::min<int>(kRifeProcesses, static_cast<int>(commands.size()));
    std::vector<std::thread> threads;
    for (int i = 0; i < workers; ++i) {
        threads.emplace_back(worker);
    }
    for (std::thread& thread : threads) {
        thread.join();
    }
    throwIfCancelled(cancelled);
    if (!failure.empty()) {
        throw std::runtime_error(failure);
    }
}

// Runs `work` over the numbers 0 to count, on `jobs` threads.
void parallelFor(std::size_t count, int jobs,
                 const std::function<void(std::size_t)>& work) {
    if (count == 0) {
        return;
    }
    std::atomic<std::size_t> next{0};
    std::mutex lock;
    std::string failure;
    auto worker = [&]() {
        while (true) {
            const std::size_t at = next.fetch_add(1);
            if (at >= count) {
                return;
            }
            try {
                work(at);
            } catch (const std::exception& error) {
                std::lock_guard<std::mutex> guard(lock);
                if (failure.empty()) {
                    failure = error.what();
                }
                return;
            }
        }
    };
    const int workers = std::max(1, std::min<int>(jobs, static_cast<int>(count)));
    std::vector<std::thread> threads;
    for (int i = 0; i < workers; ++i) {
        threads.emplace_back(worker);
    }
    for (std::thread& thread : threads) {
        thread.join();
    }
    if (!failure.empty()) {
        throw std::runtime_error(failure);
    }
}

void describe(const Json& spec, const std::vector<PlannedFrame>& frames, int fps,
              const Rect& rect) {
    std::map<std::string, int> byKind;
    for (const PlannedFrame& frame : frames) {
        byKind[frameKindName(frame.kind)] += 1;
    }
    std::cout << "capture " << spec.at("capture").get<std::string>() << ", "
              << spec.at("poses").size() << " poses, end "
              << spec.at("end_ms").get<std::int64_t>() << " ms, " << frames.size()
              << " frames at " << fps << " fps\n";
    std::cout << "canvas " << rect.width << "x" << rect.height << " at (" << rect.x << ", "
              << rect.y << ")\n";
    std::cout << "frames by kind: ";
    bool first = true;
    for (const auto& kind : byKind) {
        if (!first) {
            std::cout << ", ";
        }
        first = false;
        std::cout << kind.first << " " << kind.second;
    }
    std::cout << std::endl;
}

}  // namespace

const char* const kModelName = "rife-v4.6";

Json generate(const PipelineOptions& options, const std::atomic<bool>* cancelled) {
    const auto started = std::chrono::steady_clock::now();

    const fs::path sourceDir = fs::path(options.assetsDir) / "captures" / options.capture;
    const std::string specPath =
        options.resolvedPath.empty() ? (sourceDir / "resolved.json").string()
                                     : options.resolvedPath;
    if (!fs::exists(specPath)) {
        throw std::runtime_error("no timeline at " + specPath);
    }
    const Json spec = readJsonFile(specPath);
    const std::string specHash = sha256File(specPath);
    const std::vector<PlannedFrame> frames = planFrames(spec, options.fps);
    const Rect rect = unionRect(spec.at("poses"));

    if (options.dryRun) {
        describe(spec, frames, options.fps, rect);
        return nullptr;
    }

    // Everything lands beside the requested directory and moves into place at
    // the end, so a run that stops halfway leaves the old frames alone.
    const fs::path finalDir(options.outDir);
    const std::string stamp = std::to_string(static_cast<long>(::getpid()));
    const fs::path stagingDir =
        finalDir.parent_path() / (finalDir.filename().string() + ".new-" + stamp);
    fs::remove_all(stagingDir);
    fs::create_directories(stagingDir);
    const fs::path workDir = stagingDir / "work";

    try {
        // Compose every pose and write the two pictures RIFE reads.
        std::vector<std::vector<std::uint8_t>> composed;
        composed.reserve(spec.at("poses").size());
        for (const char* name : {"rgb", "alpha"}) {
            fs::remove_all(workDir / name);
            fs::create_directories(workDir / name);
        }
        for (std::size_t i = 0; i < spec.at("poses").size(); ++i) {
            const Json& pose = spec.at("poses")[i];
            std::vector<std::uint8_t> rgba =
                compose((sourceDir / pose.at("image").get<std::string>()).string(), pose, rect);
            std::vector<std::uint8_t> rgb;
            std::vector<std::uint8_t> alpha;
            split(rgba, &rgb, &alpha);
            const std::string name = numbered("%05lld.png", static_cast<long long>(i));
            writeFileBytes(workDir / "rgb" / name,
                           encodePng(rect.width, rect.height, kPngRgb, rgb, 1));
            writeFileBytes(workDir / "alpha" / name,
                           encodePng(rect.width, rect.height, kPngGray, alpha, 1));
            composed.push_back(std::move(rgba));
        }
        throwIfCancelled(cancelled);

        RifePlan plan = planRife(frames, workDir, options.rifeBinary, options.modelDir);
        runCommands(plan.commands, cancelled);
        throwIfCancelled(cancelled);

        // Where each pose index sits in the pose list.
        std::map<int, std::size_t> byIndex;
        for (std::size_t i = 0; i < spec.at("poses").size(); ++i) {
            byIndex[spec.at("poses")[i].at("index").get<int>()] = i;
        }
        const std::vector<std::uint8_t> empty(
            static_cast<std::size_t>(rect.width) * rect.height * 4, 0);

        std::vector<std::string> names;
        names.reserve(frames.size());
        for (std::size_t n = 0; n < frames.size(); ++n) {
            names.push_back(numbered("frame%05lld.png", static_cast<long long>(n)));
        }

        const int jobs = options.jobs > 0
                             ? options.jobs
                             : std::max<int>(1, static_cast<int>(
                                                    std::thread::hardware_concurrency()));
        parallelFor(frames.size(), jobs, [&](std::size_t n) {
            const fs::path outPath = stagingDir / names[n];
            if (frames[n].kind == FrameKind::Interp) {
                const auto& pair = plan.paths.at(n);
                Picture colour = readPng(pair.first);
                Picture alpha = readPng(pair.second);
                if (colour.width != rect.width || colour.height != rect.height ||
                    alpha.width != rect.width || alpha.height != rect.height) {
                    throw std::runtime_error("rife returned the wrong size for " + pair.first);
                }
                std::vector<std::uint8_t> rgb;
                if (colour.channels == 3) {
                    rgb = std::move(colour.pixels);
                } else {
                    rgb.resize(static_cast<std::size_t>(rect.width) * rect.height * 3);
                    for (std::size_t i = 0; i < rgb.size() / 3; ++i) {
                        for (int c = 0; c < 3; ++c) {
                            rgb[i * 3 + c] = colour.pixels[i * colour.channels + c];
                        }
                    }
                }
                std::vector<std::uint8_t> grey;
                if (alpha.channels == 1) {
                    grey = std::move(alpha.pixels);
                } else {
                    grey.resize(static_cast<std::size_t>(rect.width) * rect.height);
                    for (std::size_t i = 0; i < grey.size(); ++i) {
                        grey[i] = alpha.pixels[i * alpha.channels];
                    }
                }
                writeFileBytes(outPath, encodePng(rect.width, rect.height, kPngRgba,
                                                  combine(rgb, grey, kAlphaCutoff), 6));
            } else if (frames[n].kind == FrameKind::Blank) {
                writeFileBytes(outPath, encodePng(rect.width, rect.height, kPngRgba, empty, 6));
            } else {
                writeFileBytes(outPath,
                               encodePng(rect.width, rect.height, kPngRgba,
                                         composed[byIndex.at(frames[n].source[0])], 6));
            }
        });
        throwIfCancelled(cancelled);

        // The manifest.
        Json tool = Json::object();
        tool["commit"] = toolCommit();
        tool["model"] = kModelName;
        tool["model_sha256"] = sha256File((fs::path(options.modelDir) / "flownet.bin").string());
        tool["binary_sha256"] = sha256File(options.rifeBinary);
        tool["alpha_cutoff"] = kAlphaCutoff;
        tool["rife_commands"] = static_cast<std::uint64_t>(plan.commands.size());
        tool["rife_example"] = plan.commands.empty()
                                   ? Json(nullptr)
                                   : Json(joinArguments(plan.commands.front()));

        std::map<std::string, std::string> hashes;
        Json posesOut = Json::array();
        for (const Json& pose : spec.at("poses")) {
            const std::string image = pose.at("image").get<std::string>();
            auto found = hashes.find(image);
            if (found == hashes.end()) {
                found = hashes.emplace(image, sha256File((sourceDir / image).string())).first;
            }
            Json entry = Json::object();
            entry["index"] = pose.at("index");
            entry["t_ms"] = pose.at("t_ms");
            entry["image"] = image;
            entry["x"] = pose.at("x");
            entry["y"] = pose.at("y");
            entry["w"] = pose.at("w");
            entry["h"] = pose.at("h");
            entry["sha256"] = found->second;
            posesOut.push_back(entry);
        }

        Json input = Json::object();
        input["resolved"] = specPath;
        input["resolved_sha256"] = specHash;
        input["source_dir"] = sourceDir.string();
        input["from_fixture"] = false;
        input["poses"] = posesOut;

        Json frameList = Json::array();
        const Fraction end(spec.at("end_ms").get<std::int64_t>());
        std::vector<Fraction> starts;
        starts.reserve(frames.size());
        for (const PlannedFrame& frame : frames) {
            starts.push_back(frame.t.roundTo6());
        }
        for (std::size_t n = 0; n < frames.size(); ++n) {
            const Fraction stop = n + 1 < starts.size() ? starts[n + 1] : end;
            Json entry = Json::object();
            entry["file"] = names[n];
            entry["t_ms"] = starts[n].toDouble();
            entry["duration_ms"] = (stop - starts[n]).toDouble();
            entry["kind"] = frameKindName(frames[n].kind);
            Json source = Json::array();
            for (int index : frames[n].source) {
                source.push_back(index);
            }
            entry["source"] = source;
            entry["s"] = frames[n].hasStep ? Json(frames[n].step.toDouble()) : Json(nullptr);
            frameList.push_back(entry);
        }

        Json manifest = Json::object();
        manifest["capture"] = options.capture;
        manifest["fps"] = options.fps;
        manifest["tool"] = tool;
        manifest["input"] = input;
        manifest["canvas"] = spec.at("canvas");
        Json frameRect = Json::object();
        frameRect["x"] = rect.x;
        frameRect["y"] = rect.y;
        frameRect["w"] = rect.width;
        frameRect["h"] = rect.height;
        manifest["frame_rect"] = frameRect;
        manifest["end_ms"] = spec.at("end_ms");
        manifest["final_wav"] = spec.contains("final_wav") ? spec.at("final_wav") : Json(nullptr);
        manifest["cuts"] = spec.contains("cuts") && !spec.at("cuts").is_null()
                               ? spec.at("cuts")
                               : Json::array();
        manifest["pre_sounds"] = spec.contains("pre_sounds") && !spec.at("pre_sounds").is_null()
                                     ? spec.at("pre_sounds")
                                     : Json::array();
        manifest["sounds"] = soundEvents(spec);
        manifest["frames"] = frameList;

        const double seconds =
            std::chrono::duration<double>(std::chrono::steady_clock::now() - started).count();
        char rounded[32];
        std::snprintf(rounded, sizeof(rounded), "%.2f", seconds);
        manifest["wall_seconds"] = std::strtod(rounded, nullptr);
        swchess::exporter::writeJsonFile((stagingDir / "manifest.json").string(), manifest, true,
                                         true);

        if (!options.keepWork) {
            fs::remove_all(workDir);
        }

        const fs::path retired =
            finalDir.parent_path() / (finalDir.filename().string() + ".old-" + stamp);
        if (fs::exists(finalDir)) {
            fs::rename(finalDir, retired);
        }
        fs::create_directories(finalDir.parent_path());
        fs::rename(stagingDir, finalDir);
        fs::remove_all(retired);
        return manifest;
    } catch (...) {
        fs::remove_all(stagingDir);
        throw;
    }
}

bool refreshCues(const std::string& outDir, const std::string& resolvedPath) {
    const fs::path manifestPath = fs::path(outDir) / "manifest.json";
    Json manifest = readJsonFile(manifestPath);
    std::string path = resolvedPath;
    if (path.empty() && manifest.contains("input") && manifest["input"].contains("resolved") &&
        !manifest["input"]["resolved"].is_null()) {
        path = manifest["input"]["resolved"].get<std::string>();
    }
    if (path.empty() || !fs::exists(path)) {
        throw std::runtime_error(manifestPath.string() + " names no resolved.json to read cues from");
    }
    const Json spec = readJsonFile(path);

    const std::string before =
        swchess::exporter::dumpPythonJson(manifest["sounds"], true) +
        swchess::exporter::dumpPythonJson(manifest["pre_sounds"], true) +
        swchess::exporter::dumpPythonJson(manifest["final_wav"], true);

    manifest["sounds"] = soundEvents(spec);
    manifest["pre_sounds"] = spec.contains("pre_sounds") && !spec.at("pre_sounds").is_null()
                                 ? spec.at("pre_sounds")
                                 : Json::array();
    manifest["final_wav"] = spec.contains("final_wav") ? spec.at("final_wav") : Json(nullptr);
    manifest["input"]["resolved"] = path;
    manifest["input"]["resolved_sha256"] = sha256File(path);

    const std::string after =
        swchess::exporter::dumpPythonJson(manifest["sounds"], true) +
        swchess::exporter::dumpPythonJson(manifest["pre_sounds"], true) +
        swchess::exporter::dumpPythonJson(manifest["final_wav"], true);

    swchess::exporter::writeJsonFile(manifestPath.string(), manifest, true, true);
    return before != after;
}

}  // namespace swchess::interp
